Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm

```diff
@@ -605,6 +605,107 @@ - (void)setProxyConfigurations:(NSArray<nw_proxy_config_t> *)proxyConfigurations
     return *_websiteDataStore;
 }
 
+struct WKWebsiteData {
+    std::optional<HashMap<WebCore::ClientOrigin, HashMap<String, String>>> localStorage;
+};
+
+- (void)fetchDataOfTypes:(NSSet<NSString *> *)dataTypes completionHandler:(void(^)(NSData *, NSError *))completionHandler
+{
+    Vector<WebKit::WebsiteDataType> dataTypesToEncode;
+    for (NSString *dataType in dataTypes) {
+        if ([dataType isEqualToString:WKWebsiteDataTypeLocalStorage]) {
+            dataTypesToEncode.append(WebKit::WebsiteDataType::LocalStorage);
+            continue;
+        }
+
+        completionHandler(NULL, [[NSError alloc] initWithDomain:WKErrorDomain code:NSFeatureUnsupportedError userInfo:NULL]);
+    }
+
+    auto data = Box<WKWebsiteData>::create();
+
+    auto callbackAggregator = CallbackAggregator::create([completionHandler = makeBlockPtr(completionHandler), dataTypesToEncode = WTFMove(dataTypesToEncode), data] {
+        WTF::Persistence::Encoder encoder;
+        constexpr unsigned currentWKWebsiteDataSerializationVersion = 1;
+        encoder << currentWKWebsiteDataSerializationVersion;
+        encoder << dataTypesToEncode;
+
+        for (auto& dataTypeToEncode : dataTypesToEncode) {
+            switch (dataTypeToEncode) {
+            case WebKit::WebsiteDataType::LocalStorage:
+                if (!data->localStorage) {
+                    completionHandler(NULL, [[NSError alloc] initWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:NULL]);
```

Comment:
I'm sure this is not the right error but am not sure what would be better. This will be called if the Network Process crashes and the completion handler is called with the default arguments (std::nullopt). Any suggestions?